File: tests/test_crop.py
"""Tests for the crop operation (bounding box / half-space)."""

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus import crop


def _quad_grid():
    # 2x1 grid of quads (3x2 points), x in [0,2], y in [0,1]
    pts = np.array(
        [[0, 0, 0], [1, 0, 0], [2, 0, 0], [0, 1, 0], [1, 1, 0], [2, 1, 0]], float
    )
    return meshioplusplus.Mesh(pts, [("quad", np.array([[0, 1, 4, 3], [1, 2, 5, 4]]))])


def test_bbox_keeps_cells_inside():
    mesh = _quad_grid()
    out = crop(mesh, bbox=[-0.1, -0.1, -0.1, 1.1, 1.1, 1.1], mode="all")
    assert len(out.cells[0].data) == 1  # only the first quad is fully inside
    # no unused points
    used = set(out.cells[0].data.reshape(-1).tolist())
    assert used == set(range(len(out.points)))


def test_mode_all_vs_any_on_straddling_cell():
    mesh = _quad_grid()
    box = [-0.1, -0.1, -0.1, 1.1, 1.1, 1.1]
    all_mode = crop(mesh, bbox=box, mode="all")
    any_mode = crop(mesh, bbox=box, mode="any")
    assert len(all_mode.cells[0].data) == 1
    assert len(any_mode.cells[0].data) == 2  # second quad straddles the boundary


def test_halfspace():
    mesh = _quad_grid()
    # keep x >= 1  (normal +x, point at x=1)
    out = crop(mesh, plane=([1, 0, 0], [1, 0, 0]), mode="all")
    assert len(out.cells[0].data) == 1  # only the right quad


def test_record_ids():
    mesh = _quad_grid()
    out = crop(
        mesh, bbox=[-0.1, -0.1, -0.1, 1.1, 1.1, 1.1], mode="all", record_ids=True
    )
    assert "crop:original_point_id" in out.point_data
    assert "crop:original_cell_id" in out.cell_data
    # original point ids are a subset of the input point indices
    assert set(out.point_data["crop:original_point_id"].tolist()).issubset(
        set(range(6))
    )


def test_pruned_mesh_has_no_unused_points():
    mesh = _quad_grid()
    out = crop(mesh, bbox=[-0.1, -0.1, -0.1, 1.1, 1.1, 1.1], mode="all")
    used = np.zeros(len(out.points), dtype=bool)
    for cb in out.cells:
        used[cb.data.reshape(-1)] = True
    assert used.all()


def test_cpp_matches_python():
    core = pytest.importorskip("meshioplusplus._core")
    mesh = _quad_grid()
    got = core.crop_bbox(mesh, [-0.1, -0.1, -0.1], [1.1, 1.1, 1.1], "any", False)
    # against the numpy fallback via the public API (which uses _core) — compare
    # point counts / cell counts as a structural check
    from meshioplusplus._crop import _crop_py, _points_3d

    p3 = _points_3d(mesh)
    mask = np.all((p3 >= [-0.1, -0.1, -0.1]) & (p3 <= [1.1, 1.1, 1.1]), axis=1)
    ref, _, _ = _crop_py(mesh, mask, "any", False)
    assert len(got["mesh"].points) == len(ref.points)
    assert len(got["mesh"].cells[0].data) == len(ref.cells[0].data)


def test_roundtrip_write_read(tmp_path):
    mesh = _quad_grid()
    out = crop(mesh, bbox=[-0.1, -0.1, -0.1, 1.1, 1.1, 1.1], mode="all")
    p = tmp_path / "cr.vtu"
    meshioplusplus.write(p, out)
    back = meshioplusplus.read(p)
    assert len(back.points) == len(out.points)


def test_requires_exactly_one_region():
    mesh = _quad_grid()
    with pytest.raises(ValueError):
        crop(mesh)
    with pytest.raises(ValueError):
        crop(mesh, bbox=[0, 0, 0, 1, 1, 1], plane=([0, 0, 0], [1, 0, 0]))

File: tests/cpp/test_crop.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// Tests for the crop operation (bounding box / half-space / cell_data predicate).

// System includes
#include <cmath>
#include <limits>
#include <stdexcept>

// External includes
#include <gtest/gtest.h>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/detail/value_io.hpp"
#include "meshioplusplus/operations/crop.hpp"

namespace {

using meshioplusplus::crop_bbox;
using meshioplusplus::crop_halfspace;
using meshioplusplus::crop_predicate;
using meshioplusplus::CropMode;
using meshioplusplus::CropResult;
using meshioplusplus::Mesh;
using meshioplusplus::NDArray;
using meshioplusplus::RefineCompare;

// 2x1 grid of quads (points x in [0,2], y in [0,1]).
Mesh quad_grid() {
    return mt::make_mesh({{0, 0, 0}, {1, 0, 0}, {2, 0, 0}, {0, 1, 0}, {1, 1, 0}, {2, 1, 0}}, "quad",
                         {{0, 1, 4, 3}, {1, 2, 5, 4}});
}

std::size_t count_used(const Mesh& m) {
    std::size_t used = 0;
    std::vector<char> seen(m.NumPoints(), 0);
    for (const auto cb : m.CellRange()) {
        const auto& conn = cb.Conn();
        for (std::size_t i = 0; i < conn.Size(); ++i)
            seen[static_cast<std::size_t>(conn.As<std::int64_t>()[i])] = 1;
    }
    for (char c : seen)
        used += c ? 1 : 0;
    return used;
}

TEST(Crop, BBoxModeAll) {
    Mesh m = quad_grid();
    double lo[3] = {-0.1, -0.1, -0.1}, hi[3] = {1.1, 1.1, 1.1};
    CropResult r = crop_bbox(m, lo, hi, CropMode::All);
    ASSERT_EQ(r.mMesh.NumCellBlocks(), 1u);
    EXPECT_EQ(r.mMesh.Cells(0).NumCells(), 1u);           // only the fully-inside quad
    EXPECT_EQ(count_used(r.mMesh), r.mMesh.NumPoints());  // no unused points
}

TEST(Crop, BBoxModeAnyKeepsStraddling) {
    Mesh m = quad_grid();
    double lo[3] = {-0.1, -0.1, -0.1}, hi[3] = {1.1, 1.1, 1.1};
    CropResult r = crop_bbox(m, lo, hi, CropMode::Any);
    EXPECT_EQ(r.mMesh.Cells(0).NumCells(), 2u);
}

TEST(Crop, HalfSpace) {
    Mesh m = quad_grid();
    double point[3] = {1, 0, 0}, normal[3] = {1, 0, 0};  // keep x >= 1
    CropResult r = crop_halfspace(m, point, normal, CropMode::All);
    EXPECT_EQ(r.mMesh.Cells(0).NumCells(), 1u);
}

// A quad grid whose cells carry a scalar `t` = 0, 1, ... and, when asked, a NaN.
Mesh tagged_quad_grid(bool WithNan) {
    Mesh m = mt::make_mesh(
        {{0, 0, 0}, {1, 0, 0}, {2, 0, 0}, {3, 0, 0}, {0, 1, 0}, {1, 1, 0}, {2, 1, 0}, {3, 1, 0}},
        "quad", {{0, 1, 5, 4}, {1, 2, 6, 5}, {2, 3, 7, 6}});
    NDArray t = NDArray::Uninit(meshioplusplus::DType::Float64, {3u});
    t.As<double>()[0] = 0.0;
    t.As<double>()[1] = WithNan ? std::numeric_limits<double>::quiet_NaN() : 1.0;
    t.As<double>()[2] = 2.0;
    std::vector<NDArray> blocks;
    blocks.push_back(std::move(t));
    m.AddCellData("t", std::move(blocks));
    return m;
}

TEST(Crop, PredicateKeepsTheMatchingCells) {
    const CropResult r = crop_predicate(tagged_quad_grid(false), "t", RefineCompare::Less, 1.5);
    ASSERT_EQ(r.mMesh.NumCellBlocks(), 1u);
    EXPECT_EQ(r.mMesh.Cells(0).NumCells(), 2u);
    // Points the dropped cell alone referenced are pruned, exactly as the
    // spatial crops do -- all three funnel through the same subset builder.
    EXPECT_EQ(count_used(r.mMesh), r.mMesh.NumPoints());
    EXPECT_EQ(r.mMesh.NumPoints(), 6u);
    // The predicate array itself rides through, subsetted.
    ASSERT_TRUE(r.mMesh.HasCellData("t"));
    EXPECT_DOUBLE_EQ(meshioplusplus::detail::read_double(r.mMesh.CellData("t", 0), 0), 0.0);
    EXPECT_DOUBLE_EQ(meshioplusplus::detail::read_double(r.mMesh.CellData("t", 0), 1), 1.0);
}

TEST(Crop, PredicateHonoursEveryComparison) {
    const auto n = [](RefineCompare op, double v) {
        return crop_predicate(tagged_quad_grid(false), "t", op, v).mMesh.Cells(0).NumCells();
    };
    EXPECT_EQ(n(RefineCompare::Less, 1.0), 1u);
    EXPECT_EQ(n(RefineCompare::LessEqual, 1.0), 2u);
    EXPECT_EQ(n(RefineCompare::Greater, 1.0), 1u);
    EXPECT_EQ(n(RefineCompare::GreaterEqual, 1.0), 2u);
    EXPECT_EQ(n(RefineCompare::Equal, 1.0), 1u);
    EXPECT_EQ(n(RefineCompare::NotEqual, 1.0), 2u);
}

// The rule `refine` already states and this operation inherits by sharing the
// evaluator: a non-finite value never matches, whatever the comparison. It has
// to hold for `!=` too, which is the case a naive implementation gets wrong --
// `NaN != 1.0` is true in IEEE.
TEST(Crop, ANonFiniteCellValueNeverMatches) {
    for (RefineCompare op :
         {RefineCompare::Less, RefineCompare::LessEqual, RefineCompare::Greater,
          RefineCompare::GreaterEqual, RefineCompare::Equal, RefineCompare::NotEqual}) {
        const CropResult r = crop_predicate(tagged_quad_grid(true), "t", op, 1.0);
        for (std::size_t c = 0; c < r.mMesh.Cells(0).NumCells(); ++c)
            EXPECT_TRUE(
                std::isfinite(meshioplusplus::detail::read_double(r.mMesh.CellData("t", 0), c)))
                << "the NaN cell survived comparison " << static_cast<int>(op);
    }
}

TEST(Crop, PredicateRejectsWhatIsNotAScalarCellArray) {
    Mesh m = tagged_quad_grid(false);
    // An unknown name, with the available ones listed.
    EXPECT_THROW(crop_predicate(m, "nope", RefineCompare::Less, 1.0), std::invalid_argument);
    // point_data is refused BY NAME rather than averaged onto the cells.
    NDArray pd = NDArray(meshioplusplus::DType::Float64, {m.NumPoints()});
    m.AddPointData("p", std::move(pd));
    try {
        crop_predicate(m, "p", RefineCompare::Less, 1.0);
        FAIL() << "expected a throw";
    } catch (const std::invalid_argument& e) {
        EXPECT_NE(std::string(e.what()).find("point_data"), std::string::npos);
    }
    // A vector array has no single value per cell to compare.
    NDArray v = NDArray(meshioplusplus::DType::Float64, {3u, 3u});
    std::vector<NDArray> vb;
    vb.push_back(std::move(v));
    m.AddCellData("v", std::move(vb));
    EXPECT_THROW(crop_predicate(m, "v", RefineCompare::Less, 1.0), std::invalid_argument);
}

TEST(Crop, PredicateRecordsIdsLikeTheOthers) {
    const CropResult r =
        crop_predicate(tagged_quad_grid(false), "t", RefineCompare::Less, 1.5, true);
    EXPECT_TRUE(r.mMesh.HasPointData("crop:original_point_id"));
    EXPECT_TRUE(r.mMesh.HasCellData("crop:original_cell_id"));
    EXPECT_EQ(meshioplusplus::detail::read_int(r.mMesh.CellData("crop:original_cell_id", 0), 1), 1);
}

TEST(Crop, RecordIds) {
    Mesh m = quad_grid();
    double lo[3] = {-0.1, -0.1, -0.1}, hi[3] = {1.1, 1.1, 1.1};
    CropResult r = crop_bbox(m, lo, hi, CropMode::All, /*record_ids=*/true);
    EXPECT_TRUE(r.mMesh.HasPointData("crop:original_point_id"));
    EXPECT_TRUE(r.mMesh.HasCellData("crop:original_cell_id"));
}

}  // namespace
